File: android/src/main/kotlin/noise/protocol/KeyStore.kt
package noise.protocol

/**
 * Manages handshake key state with domain-error accessors.
 *
 * Replaces nullable vars + force-unwraps (`!!`) in HandshakeState with
 * explicit [requireKeyPair] / [requirePublicKey] methods that throw
 * [NoiseException.InvalidKey] with descriptive messages.
 */
class KeyStore(
    staticKeyPair: KeyPair? = null,
    remoteStaticKey: ByteArray? = null
) {
    internal var s: KeyPair? = staticKeyPair
        private set
    internal var e: KeyPair? = null
    internal var rs: ByteArray? = remoteStaticKey
    internal var re: ByteArray? = null

    /** Returns the local key pair for [ref], or throws [NoiseException.InvalidKey]. */
    fun requireKeyPair(ref: KeyRef): KeyPair = when (ref) {
        KeyRef.S -> s ?: throw NoiseException.InvalidKey("Local static key not available")
        KeyRef.E -> e ?: throw NoiseException.InvalidKey("Local ephemeral key not yet generated")
        else -> throw NoiseException.InvalidKey("$ref is not a local key pair reference")
    }

    /** Returns the public key for [ref], or throws [NoiseException.InvalidKey]. */
    fun requirePublicKey(ref: KeyRef): ByteArray = when (ref) {
        KeyRef.S -> s?.publicKey ?: throw NoiseException.InvalidKey("Local static key not available")
        KeyRef.E -> e?.publicKey ?: throw NoiseException.InvalidKey("Local ephemeral key not yet generated")
        KeyRef.RS -> rs ?: throw NoiseException.InvalidKey("Remote static key not yet received")
        KeyRef.RE -> re ?: throw NoiseException.InvalidKey("Remote ephemeral key not yet received")
    }
}

File: ios/Sources/NoiseProtocol/KeyStore.swift
import Foundation

/// Manages handshake key state with domain-error accessors.
class KeyStore {
    var s: KeyPair?
    var e: KeyPair?
    var rs: Data?
    var re: Data?

    init(staticKeyPair: KeyPair? = nil, remoteStaticKey: Data? = nil) {
        self.s = staticKeyPair
        self.rs = remoteStaticKey
    }

    /// Returns the local key pair for the given reference, or throws.
    func requireKeyPair(_ ref: KeyRef) throws -> KeyPair {
        switch ref {
        case .s:
            guard let kp = s else {
                throw NoiseError.invalidKey("Local static key not available")
            }
            return kp
        case .e:
            guard let kp = e else {
                throw NoiseError.invalidKey("Local ephemeral key not yet generated")
            }
            return kp
        default:
            throw NoiseError.invalidKey("\(ref) is not a local key pair reference")
        }
    }

    /// Returns the public key for the given reference, or throws.
    func requirePublicKey(_ ref: KeyRef) throws -> Data {
        switch ref {
        case .s:
            guard let pk = s?.publicKey else {
                throw NoiseError.invalidKey("Local static key not available")
            }
            return pk
        case .e:
            guard let pk = e?.publicKey else {
                throw NoiseError.invalidKey("Local ephemeral key not yet generated")
            }
            return pk
        case .rs:
            guard let pk = rs else {
                throw NoiseError.invalidKey("Remote static key not yet received")
            }
            return pk
        case .re:
            guard let pk = re else {
                throw NoiseError.invalidKey("Remote ephemeral key not yet received")
            }
            return pk
        }
    }
}
